compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -358,6 +359,21 @@ impl<'a, 'tcx> MatchCheckCtxt<'a, 'tcx> {
             _ => false,
         }
     }
+
+    /// Type inference occasionally gives us opaque types in places where corresponding patterns
+    /// have more specific types. To avoid inconsistencies as well as detect opaque uninhabited
+    /// types, we use the corresponding concrete type if possible.
+    fn reveal_opaque_ty(&self, ty: Ty<'tcx>) -> Ty<'tcx> {
```

Comment:
Do you only care about opaque types themselves, or also about types containing opaque types? Something like `match (42, recur()) {}` may have similar issues if `recur()` is returning a diverging, but opaque type

Reply:
Ah yes that's the subtle point I hinted at: in a case like this I rely on `Ty::inhabited_predicate` to tell me about inhabitedness, but `Ty::inhabited_predicate` never looks into opaque types. This is inconsistent: `match (42, recur()) {}` would error "type is not empty", yet
```rust
match (42, recur()) {
  (0, _) => {}
}
```
would report the arm as unreachable (if `exhaustive_patterns` is turned on or eventually stabilized).

For consistency I think `Ty::inhabited_predicate` should look into opaques, but I don't know the ramifications that could have. Plus it currently tracks the visibility of inhabitedness at a module granularity, and this would require function granularity.

Reply:
It's even more inconsistent than that: the first is accepted, and the second warns as unreachable:
```rust
match (42, recur()) {
  _ => {}
}
match (42, recur()) {
  (_, _) => {}
}
```
(This is under `exhaustive_patterns` only for now. None of these inconsistencies are observable on stable rust because exhaustiveness treats nested empty types as nonempty)

Reply:
> Alternatively you can make reveal_opaque_ty be a type folder

Oh I didn't know about `ty.has_opaque_types()`! Yeah that could work, I like that idea. With the added benefit that if I'm careful I'll only need a single `reveal_opaque_types` on the scrutinee type at the start, and then I'd be operating on fully revealed types

Reply:
I don't think it will have a perf effect in practice, we do similar operations elsewhere in the compiler regularly. So yea, please add it to this PR

Reply:
Done. Could you have a quick look at my `TypeFolder` impl? It's my first one.

> I don't think it will have a perf effect in practice

Ok :+1:. Before `inhabitedness_predicate` was a query, the checks for `is_uninhabited` all around exhaustiveness had a [significant perf impact](https://github.com/rust-lang/rust/pull/79394#issuecomment-733706939) so I'm extra careful now x)

Reply:
Ah, it explodes on recursive types https://github.com/Nadrieril/rust/blob/fix-opaque-ice/tests/ui/type-alias-impl-trait/recursive-fn-tait.rs . Do you know how I could avoid that?

Reply:
Alright I've got a solution but I'm not fully satisfied: when I detect a recursive loop, I leave the opaque as-is, i.e. unrevealed. Even though I technically have enough information to reveal it. E.g.:
```rust
type W = impl Copy;
#[derive(Copy, Clone)]
struct Rec<'a> {
    n: u32,
    w: Option<&'a W>,
}
fn recursive_opaque() -> W {
    if true {
        match recursive_opaque() {
            Rec { n: 0, w: Some(Rec { n: 0, w: _ }) } => {}
            _ => {}
        }
    }
    Rec { n: 0, w: None }
}
```

In a case like this, at the start of exhaustiveness I try to reveal the type but only reveal the first layer. This is fine currently because I call `reveal_opaque` all the time, but I hope this won't bite me later.

Reply:
Scratch that, type folding does not recurse into the fields of an ADT anyway. My solution is fine.